### ISSUE TITLE
Manipulating velocity conflicts with Multiway

### ISSUE THREAD

**Issue by mucaho** [craftyjs/Crafty#904]
As described [in this post](https://github.com/craftyjs/Crafty/issues/903#issuecomment-102533377) resetting velocity `vx = 0` while a key is pressed will lead to undesired behavior after releasing said key.
[Relevant lines](https://github.com/craftyjs/Crafty/blob/e79a84cc5a39a39603b7020192301da1944258fc/src/controls/controls.js#L158-L178):

``` js
_keydown: function (e) { ...
    this.vx += directionSpeed.x;
    this.vy += directionSpeed.y;
... },

_keyup: function (e) { ...
    this.vx -= directionSpeed.x;
    this.vy -= directionSpeed.y;
... },
```

**Comment by ashes999** [comment]
I just ran into this again today. Is there a work-around or fix? Maybe a way to specify an impulse?

**Comment by starwed** [comment]
The workaround would be to not use Multiway, but use it as a model to build your own component.

Multiway should probably be fixed so that letting go of the key will never change direction.

**Comment by mucaho** [comment]
there is no solution to this problem I can think of so far:
- snap velocity hard
  - won't reactivate movement while key is pressed after `v` was set to `0` upon collision event (as in linked code example) - e.g. player temporarily collides with asteroid, then asteroid moves away, but player doesn't move while key is still being pressed
  - won't behave correctly with diagonal movement (where keys affect both `vx` and `vy`)
  - doesn't interact nicely with other features that might affect velocity (e.g. detached entity could retain former parent's velocity in addition to its current velocity)

``` js
_keydown: function (e) { ...
    this.vx = directionSpeed.x;
    this.vy = directionSpeed.y;
... },

_keyup: function (e) { ...
    this.vx = 0;
    this.vy = 0;
... },
```
- check for current velocity in each frame and adjust it if it's different from desired velocity
  - won't behave correctly with diagonal movement (where keys affect both `vx` and `vy`)
  - doesn't interact nicely with other features that might affect velocity (e.g. detached entity could retain former parent's velocity in addition to its current velocity)

``` js
_keyup: function (e) { ...
    if (abs(directionSpeed.x) > 0) this.vx = 0;
    if (abs(directionSpeed.y) > 0) this.vy = 0;
... },

_enterframe: function() {...
   if (this.isDown(key)) {
      if (this.vx < directionSpeed.x && directionSpeed.x > 0) this.vx = directionSpeed.x;
      if (this.vx > directionSpeed.x && directionSpeed.x < 0) this.vx = directionSpeed.x;
      // analog for vy
   }
...}
```

Unless someone can come up with a solution to this problem, so far the best thing to do here imo is to explicitly state this in the API documentation (something like "if you need to resolve collisions, don't touch velocity but rather adjust position; or make sure to set velocity back after collision resolved")

**Comment by starwed** [comment]
In the systems I've played with, I had the controls set acceleration rather than velocity.  That bypasses this problem, but you do have to fuss around a bit to get the feeling right.  (You also need to worry a bit about maximum velocity, and implement some sort of friction system to slow you down when you release the keys.)

You can see what it feels like in action [here](http://ludonomicon.com/bounds/bounds.html), although since that was originally built against a couple-years old version of Crafty it doesn't take advantage of Motion at all.  (The base game is written in coffeescript, so the source might be  a little hard to follow.)

**Comment by ashes999** [comment]
@starwed I'll see if I can understand your example. Setting acceleration will work for me.

Can you please provide a simple example, or the source to your game? I don't mind CoffeeScript.

**Comment by ashes999** [comment]
I ran into this at least twice more.  I have a proposal.

Crafty, behind the scenes, applies velocity and moves an entity. Since users do want to tweak velocity themselves, the problem is that it affects how Crafty changes the velocity (eg. `FourWay` and `Gravity`).

Why not create two velocity values: a `system` velocity (which Crafty manipulates) and a `user` velocity (which users manipulate). This would allow us to affect velocity, without interfering with Crafty itself. The actual/real velocity is an aggregate/sum of these two values.

Alternatively, users can still (try to) change the system velocity value (which I did), although that leads to other problems.

What do you think @mucaho @starwed ?

**Comment by starwed** [comment]
@ashes999 I don't think that would work out well in practice.  The best solution is to make sure that the built-in components play nicely with each other, and to make it easy to handle the cases that are causing you issues.  

And also, to make it easier to work with forces/accelerations instead of velocities -- you generally need to consider both if you want a game to handle well.

**Comment by ashes999** [comment]
@starwed how would I build a grappling hook? Acceleration doesn't really fit here, it's more like movement at a fixed velocity.

My current solution is terrible (manually change the player's coordinates -- collisions don't even work properly), but acceptable because it's a prototype.

**Comment by starwed** [comment]
> @starwed how would I build a grappling hook? Acceleration doesn't really fit here, it's more like movement at a fixed velocity.

If you need to move at a fixed velocity, why don't you just set that fixed velocity?  Unless a bug is popping up somewhere, I don't see how that would conflict with Motion or Gravity (you'd have to disable gravity if the hook is supporting you, of course!)

When you use Fourway/Twoway, you're saying that you want the velocity to be directly set by the arrow keys.  If that's incompatible with some movement types (like grappling hooks) in your game, there are two basic approaches:
- Remove and add the Fourway/Twoway component when you change movement modes
- Implement your own control scheme, if you need more complete control over what's happening

(A grappling hook would almost definitely need custom controls depending on the feel of the game, so how I'd implement it would depend on how it's being used.)

**Comment by mucaho** [comment]
Ok, documentation has been updated in #938 and I think it states clearly now how Multiway works behind the scenes.
I think the current approach is the most straightforward one. Acceleration and dampening seems a bit overkill to me. However, if anyone else wants to take a swing at it, feel free to do so :)

**Comment by starwed** [comment]
Multiway now works a little differently under the hood, directly setting the velocity rather than adding/subtracting from the current values.  And the actual input logic was shunted to a control system, so it should be much easier to implement e.g. a component which sets acceleration instead of velocity.

Based on those improvements, closing this issue.
